spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/integration/outbound/PubSubMessageHandler.java

```diff
@@ -16,47 +16,28 @@
 
 package org.springframework.cloud.gcp.pubsub.integration.outbound;
 
-import com.google.cloud.pubsub.spi.v1.Publisher;
-import com.google.pubsub.v1.PubsubMessage;
-import com.google.pubsub.v1.TopicName;
 import java.io.IOException;
-import org.springframework.cloud.gcp.pubsub.integration.converters.SimpleMessageConverter;
+import org.springframework.cloud.gcp.pubsub.core.PubSubTemplate;
 import org.springframework.integration.handler.AbstractMessageHandler;
 import org.springframework.messaging.Message;
-import org.springframework.messaging.converter.MessageConversionException;
-import org.springframework.messaging.converter.MessageConverter;
 
 /**
- * Converts from internal Spring message to GCP Pub/Sub message and publishes the Pub/Sub
- * message to a Pub/Sub topic.
+ * Sends messages to Google Cloud Pub/Sub by delegating to {@link PubSubTemplate}.
  *
  * @author João André Martins
  */
 public class PubSubMessageHandler extends AbstractMessageHandler {
 
-  private Publisher publisher;
-  private MessageConverter messageConverter;
+  private PubSubTemplate pubsubTemplate;
+  private String topic;
 
-  public PubSubMessageHandler(String projectId, String topicName) throws IOException {
-    this(projectId, topicName, new SimpleMessageConverter());
-  }
-
-  public PubSubMessageHandler(String projectId, String topicName,
-      MessageConverter messageConverter) throws IOException {
-    publisher = Publisher.defaultBuilder(TopicName.create(
-        projectId, topicName)).build();
-    this.messageConverter = messageConverter;
+  public PubSubMessageHandler(String projectId, String topic) throws IOException {
+    pubsubTemplate = new PubSubTemplate(null, projectId);
+    this.topic = topic;
```

Comment:
The `PubSubTemplate.send()` should be extended to more APIs. See `KafkaOperations` for example: https://github.com/spring-projects/spring-kafka/blob/master/spring-kafka/src/main/java/org/springframework/kafka/core/KafkaOperations.java#L92.

Yes, we really can have there `PubSubMessage`-based variant as well.

Forget a bit about that `messagingGateway`. Even `@Controller` can be replaced by the Spring Integration HTTP Inbound Channel Adapter: http://docs.spring.io/spring-integration/reference/html/http.html#http-inbound

Also consider the case:
we poll data from the RDBMS and each record has its' own option about the target PubSub topic. We just store that info in the header after splitting the list of records.
Another situation that we may determine the target topic depending on the system state or some other external influence. That's how the SpEL expression (or Java `Function`) against each message comes to the rescue.

Again. Kafka is the best sample for us here: https://github.com/spring-projects/spring-integration-kafka/blob/master/src/main/java/org/springframework/integration/kafka/outbound/KafkaProducerMessageHandler.java#L147